plugins/node/opentelemetry-hapi-instrumentation/src/hapi.ts

```diff
@@ -0,0 +1,373 @@
+    // also be supported and instrumented. This is an issue with the DefinitelyTyped repo.
+    // Function is defined at: https://github.com/hapijs/hapi/blob/master/lib/index.js#L9
+    this._logger.debug('Patching Hapi.Server');
+    shimmer.wrap(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      this._moduleExports as any,
+      'Server',
+      this._getServerPatch.bind(this)
+    );
+
+    return this._moduleExports;
+  }
+
+  /**
+   * Unpatches all Hapi operations
+   */
+  protected unpatch(): void {
+    this._logger.debug('Unpatching Hapi');
+    shimmer.massUnwrap([this._moduleExports], ['server', 'Server']);
+  }
+
+  /**
+   * Patches the Hapi.server and Hapi.Server functions in order to instrument
+   * the server.route, server.ext, and server.register functions via calls to the
+   * @function _getServerRoutePatch, @function _getServerExtPatch, and
+   * @function _getServerRegisterPatch functions
+   * @param original - the original Hapi Server creation function
+   */
+  private _getServerPatch(
+    original: (options?: Hapi.ServerOptions) => Hapi.Server
+  ) {
+    const instrumentation = this;
+    return function server(this: Hapi.Server, opts?: Hapi.ServerOptions) {
+      const newServer: Hapi.Server = original.apply(this, [opts]);
+
+      shimmer.wrap(newServer, 'route', original => {
+        return instrumentation._getServerRoutePatch.bind(instrumentation)(
+          original
+        );
+      });
+
+      shimmer.wrap(newServer, 'ext', original => {
+        return instrumentation._getServerExtPatch.bind(instrumentation)(
+          original
+        );
+      });
+
+      // Casting as any is necessary here due to multiple overloads on the Hapi.Server.register
+      // function, which requires supporting a variety of different types of Plugin inputs
+      shimmer.wrap(
+        newServer,
+        'register',
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        instrumentation._getServerRegisterPatch.bind(instrumentation) as any
+      );
+      return newServer;
+    };
+  }
+
+  /**
+   * Patches the plugin register function used by the Hapi Server. This function
+   * goes through each plugin that is being registered and adds instrumentation
+   * via a call to the @function _wrapRegisterHandler function.
+   * @param {RegisterFunction<T>} original - the original register function which
+   * registers each plugin on the server
+   */
+  private _getServerRegisterPatch<T>(
+    original: RegisterFunction<T>
+  ): RegisterFunction<T> {
+    const instrumentation = this;
+    this._logger.debug('Patching Hapi.Server register function');
+    return function register(
+      this: Hapi.Server,
+      pluginInput: HapiPluginInput<T>,
+      options?: Hapi.ServerRegisterOptions
+    ) {
+      if (Array.isArray(pluginInput)) {
+        for (let i = 0; i < pluginInput.length; i++) {
```

Comment:
Could this be a for (... of ...) loop instead.

Reply:
Will do!